### PR TITLE
feat(examples/common_vmseries_and_autoscale): example for SWFW Hub

### PR DESCRIPTION
**Pull request by alperenkose**
## Description

Common VMSeries with Autoscaling example for SWFW Hub

## How Has This Been Tested?

Deployed with spoke vnet and tested inbound/outbound/east-west traffic flows.

## Types of changes

- New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes if appropriate.
- [x] All new and existing tests passed.

### CODE REVIEW

File: examples/common_vmseries_and_autoscale/variables.tf [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+  - `scalein_config` : (map|`{}`) scale in configuration, same properties supported as for `scaleout_config`
+
+  Example, no auto scaling:
+
+  ```
+  {
+  "vmss" = {
+    name              = "ngfw-vmss"
+    vnet_key          = "transit"
+    bootstrap_options = "type=dhcp-client"
```
Comment (FoSix) [inline examples/common_vmseries_and_autoscale/variables.tf:305]: there's a bug in the dedicated vmss: `"type=dhcp"`, maybe we could use this PR to fix it there

File: examples/common_vmseries_and_autoscale/example.tfvars [hunk 2]
```diff
@@ -0,0 +1,247 @@
+# --- GENERAL --- #
+location            = "North Europe"
+resource_group_name = "autoscale-common"
+name_prefix         = "example-"
+tags = {
+  "CreatedBy"   = "Palo Alto Networks"
+  "CreatedWith" = "Terraform"
+}
+enable_zones = false
```
Comment (FoSix) [inline examples/common_vmseries_and_autoscale/example.tfvars:9]: dedicated vmss is non-zonal. Maybe we could make this example zone-enabled?

**Review by FoSix — APPROVED** [review]
👍🏻 very nice, already mergeable in this form

File: examples/common_vmseries_and_autoscale/variables.tf (hunk 1 above)
Comment (alperenkose) [reply examples/common_vmseries_and_autoscale/variables.tf:305]: Yes saw that, I was going to open a separate PR for that but let me update this then.

File: examples/common_vmseries_and_autoscale/main.tf [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+  location            = var.location
+  subnet_id           = module.vnet[each.value.vnet_key].subnet_ids[each.value.subnet_key]
+
+  managed_identities = try(each.value.managed_identities, null)
+  waf_enabled        = try(each.value.waf_enabled, false)
+  capacity           = try(each.value.capacity, null)
+  capacity_min       = try(each.value.capacity_min, null)
+  capacity_max       = try(each.value.capacity_max, null)
+  enable_http2       = try(each.value.enable_http2, null)
+  zones              = var.enable_zones ? try(each.value.zones, null) : null
```
Comment (alperenkose) [inline examples/common_vmseries_and_autoscale/main.tf:155]: Shall we add this `var.enable_zones` condition to other examples as well? It doesn't check for the enable_zones variable in others.

File: examples/common_vmseries_and_autoscale/example.tfvars (hunk 2 above)
Comment (alperenkose) [reply examples/common_vmseries_and_autoscale/example.tfvars]: converted the example to zone-enabled but there is a couple of things to note; since NATGWs are zonal resources it's removed from the example and the outbound internet traffic is provided from public IP addresses assigned to public interfaces of firewalls in the VMSS. 
Another point is since we don't have a NATGW, I needed to enable public IPs on management interfaces as well, otherwise it wouldn't be possible for the firewalls to bootstrap with publicly accessible Panorama or do their initial licensing with Palo Alto update servers since they won't have internet access. 
We might add a note for this on README perhaps; mentioning that public IP addresses on management interfaces can be removed after the deployment and all management traffic can be forwarded to the firewalls' private interfaces to get NATed to the internet (after the firewalls get configured to NAT this traffic). Any other ideas?

Comment (FoSix) [reply examples/common_vmseries_and_autoscale/example.tfvars]: > Another point is since we don't have a NATGW, I needed to enable public IPs on management interfaces as well, otherwise it wouldn't be possible for the firewalls to bootstrap with publicly accessible Panorama or do their initial licensing with Palo Alto update servers since they won't have internet access.

that's not entirely true. A VM with just private IP and default NSG rules (or no NSG) can access the internet, it's not accessible from the internet. So you don't need public IPs on management to contact a publicly available Panorama (which probably will never happen) or to download updates.
In our examples we do not block the default outbound route, so we do not need public IPs for that.

File: examples/common_vmseries_and_autoscale/main.tf (hunk 3 above)
Comment (FoSix) [reply examples/common_vmseries_and_autoscale/main.tf:155]: I would keep keep it as it is and start to create a list of things to refactor. We already have #280 and #279 that will make a huge change in the code. Maybe we can move all three to a single issue?

File: examples/common_vmseries/main.tf [hunk 4]
```diff
@@ -115,7 +115,6 @@ module "load_balancer" {
       subnet_id                = try(module.vnet[v.vnet_key].subnet_ids[v.subnet_key], null)
       in_rules                 = try(v.in_rules, {})
       out_rules                = try(v.out_rules, {})
-      zones                    = var.enable_zones ? try(v.zones, null) : null # For the regions without AZ support.
```
Comment (FoSix) [inline examples/common_vmseries/main.tf:118]: why did you remove it?

File: examples/common_vmseries_and_autoscale/example.tfvars [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+    name              = "common-vmss"
+    vnet_key          = "transit"
+    zones             = ["1", "2", "3"]
+    bootstrap_options = "type=dhcp-client"
+
+    interfaces = [
+      {
+        name       = "management"
+        subnet_key = "management"
+        create_pip = true
```
Comment (FoSix) [inline examples/common_vmseries_and_autoscale/example.tfvars:205]: as discussed in some other place, we do not need this

File: examples/common_vmseries_and_autoscale/example.tfvars [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+        name              = "private"
+        subnet_key        = "private"
+        load_balancer_key = "private"
+      },
+      {
+        name                    = "public"
+        subnet_key              = "public"
+        load_balancer_key       = "public"
+        application_gateway_key = "public"
+        create_pip              = true
```
Comment (FoSix) [inline examples/common_vmseries_and_autoscale/example.tfvars:217]: we could also skip this and use outbound rules in public LB?

File: examples/dedicated_vmseries/main.tf [hunk 7]
```diff
@@ -115,7 +115,6 @@ module "load_balancer" {
       subnet_id                = try(module.vnet[v.vnet_key].subnet_ids[v.subnet_key], null)
       in_rules                 = try(v.in_rules, {})
       out_rules                = try(v.out_rules, {})
-      zones                    = var.enable_zones ? try(v.zones, null) : null # For the regions without AZ support.
```
Comment (FoSix) [inline examples/dedicated_vmseries/main.tf:118]: same here, why this got removed?

File: examples/dedicated_vmseries_and_autoscale/main.tf [hunk 8]
```diff
@@ -110,7 +110,6 @@ module "load_balancer" {
       subnet_id                = try(module.vnet[v.vnet_key].subnet_ids[v.subnet_key], null)
       in_rules                 = try(v.in_rules, {})
       out_rules                = try(v.out_rules, {})
-      zones                    = var.enable_zones ? try(v.zones, null) : null # For the regions without AZ support.
```
Comment (FoSix) [inline examples/dedicated_vmseries_and_autoscale/main.tf:113]: and again

File: examples/standalone_vmseries/main.tf [hunk 9]
```diff
@@ -115,7 +115,6 @@ module "load_balancer" {
       subnet_id                = try(module.vnet[v.vnet_key].subnet_ids[v.subnet_key], null)
       in_rules                 = try(v.in_rules, {})
       out_rules                = try(v.out_rules, {})
-      zones                    = var.enable_zones ? try(v.zones, null) : null # For the regions without AZ support.
```
Comment (FoSix) [inline examples/standalone_vmseries/main.tf:118]: same

**Review by FoSix — CHANGES_REQUESTED** [review]

File: examples/dedicated_vmseries/main.tf (hunk 7 above)
Comment (alperenkose) [reply examples/dedicated_vmseries/main.tf:118]: This variable is not there anymore in load balancer module. avzones variable within the load balancer is used instead.

File: examples/dedicated_vmseries_and_autoscale/main.tf (hunk 8 above)
Comment (alperenkose) [reply examples/dedicated_vmseries_and_autoscale/main.tf:113]: Same with others.

File: examples/common_vmseries_and_autoscale/example.tfvars (hunk 6 above)
Comment (alperenkose) [reply examples/common_vmseries_and_autoscale/example.tfvars:217]: As we don't have LB outbound rules in reference architectures we decided to keep it the same.

File: examples/common_vmseries_and_autoscale/example.tfvars (hunk 2 above)
Comment (alperenkose) [reply examples/common_vmseries_and_autoscale/example.tfvars]: After testing and troubleshooting with public IPs removed from management interfaces, we have found out that default outbound access is blocked if VMs are attached to a Load Balancer or they have any additional interface with a public IP assigned. We have both so relying on the default outbound access won't be possible. 

And since NATGWs are zonal, this leaves us with 2 feasible options I believe:
* Assigning public IP to both public and management interfaces
* Creating a separate backend pool for the management interfaces in order to create outbound rules from them (but this is problematic as well since we don’t support creating separate backend pools for different interfaces on our current LB module)

As a note if we assign public IPs to the management interfaces and the user wants to remove them we can put a comment on README saying that public IP addresses on management interfaces can be removed after the deployment with management traffic forwarded to the firewalls’ private interfaces (private LB) to get NATed to the internet (after the firewalls get configured to NAT this traffic).

Comment (alperenkose) [reply examples/common_vmseries_and_autoscale/example.tfvars]: It is agreed to keep public IP addresses on management interfaces for the simplicity of the example and put a disclaimer on how to avoid it.

File: examples/common_vmseries_and_autoscale/main.tf (hunk 3 above)
Comment (alperenkose) [reply examples/common_vmseries_and_autoscale/main.tf]: Converted it back to match with the other examples. Put a comment on #280 for this.

File: examples/common_vmseries_and_autoscale/example.tfvars (hunk 5 above)
Comment (alperenkose) [reply examples/common_vmseries_and_autoscale/example.tfvars]: Agreed to keep it. Details on the other thread.

**Review by FoSix — APPROVED** [review]

**Review by migara — APPROVED** [review]
